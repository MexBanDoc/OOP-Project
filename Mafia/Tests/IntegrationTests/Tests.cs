using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text;
using System.Threading.Tasks;
using ApprovalTests;
using ApprovalTests.Reporters;
using FluentAssertions;
using Mafia.Domain;
using NUnit.Framework;

namespace Tests.IntegrationTests
{
    [TestFixture]
    public class Tests
    {
        private class StringInterface : IUserInterface
        {
            private Dictionary<Role, IPerson> targets;
            private StringBuilder result = new StringBuilder();
            
            public Task AskForInteractionTarget(IEnumerable<IPerson> players, Role role, ICity city)
            {
                result.Append(role.DayTime == DayTime.Night ? "Город засыпает" : "Город просыпается");
                result.Append("\n");

                targets = new Dictionary<Role, IPerson>();
                
                var victims = new List<IPerson>();
                foreach (var player in players)
                {
                    result.Append($"Просыпается {role.Name} {player.Name}");
                    result.Append("\n");
                    var name = city.Population
                        .Where(person => person.NightRole is null || !person.NightRole.Equals(role))
                        .First(person => person.IsAlive).Name;
                    victims.Add(city.GetPersonByName(name));
                }

                if (victims.Count == 0)
                    targets[role] = null;
                else
                    targets[role] = victims[0];
                
                return Task.CompletedTask;
            }

            public Task<IPerson> GetInteractionTarget(Role role, ICity city)
            {
                return Task.FromResult(targets.ContainsKey(role) ? targets[role] : null);
            }

            public Task TellResults(ICity city, DayTime dayTime)
            {
                foreach (var pair in city.LastChanges)
                    result.Append($"{pair.Key.Name} {pair.Value}");
                result.Append("\n");
                result.Append("\n");
                return Task.CompletedTask;
            }

            public Task TellGameResult(WinState state, ICity city)
            {
                switch (state)
                {
                    case WinState.MafiaWins:
                        result.Append("Мафия победила");
                        break;
                    case WinState.InProcess:
                        result.Append("Ничья");
                        break;
                    case WinState.PeacefulWins:
                        result.Append("Мирные победили");
                        break;
                    case WinState.PsychoWins:
                        result.Append("Психи захватили мир!1!");
                        break;
                    default:
                        throw new ArgumentOutOfRangeException(nameof(state), state, null);
                }
                result.Append("\n");
                return Task.CompletedTask;
            }

            public string GetGameProcess() => result.ToString();
        }
        
        
        [Test]
        [MethodImpl(MethodImplOptions.NoInlining)]
        [UseReporter(typeof(DiffReporter))]
        public void PeacefulWins()
        {
            var settings = new Settings(Settings.DefaultWinCondition,
                new Dictionary<Role, int> {[new MafiaRole()] = 10}, 0);
            var city = new City(settings.GeneratePopulation(new[]
                        {
                            "Timofey",
                            "ЛенОЧКА",
                            "Android",
                            "Дыня",
                            "God",
                            "Степан",
                            "Николай"
                        },
                        new Random(156))
                    .ToHashSet(),
               settings.CityName);
            var inter = new StringInterface();
            var game = new Game(settings, city, inter);
            game.StartGame().Wait();
            game.GetGameStatus().Should().Be(WinState.PeacefulWins);
            Approvals.Verify(inter.GetGameProcess());
        }
        
        [Test]
        [MethodImpl(MethodImplOptions.NoInlining)]
        [UseReporter(typeof(DiffReporter))]
        public void MafiaWins()
        {
            var settings = new Settings(Settings.DefaultWinCondition,
                new Dictionary<Role, int> {[new MafiaRole()] = 60}, 0);
            var city = new City(settings.GeneratePopulation(new[]
                        {
                            "Timofey",
                            "Леночка",
                            "Android",
                            "Дыня",
                            "God",
                            "Степан"
                        },
                        new Random(4))
                    .ToHashSet(),
                settings.CityName);
            var inter = new StringInterface();
            var game = new Game(settings, city, inter);
            game.StartGame().Wait();
            game.GetGameStatus().Should().Be(WinState.MafiaWins);
            Approvals.Verify(inter.GetGameProcess());
        }
        
        [Test]
        [MethodImpl(MethodImplOptions.NoInlining)]
        [UseReporter(typeof(DiffReporter))]
        public void PsychoWins()
        {
            var settings = new Settings(Settings.SaneWinCondition, Settings.CrazyNosyBizarreTown.PlayerDistribution,
                0);
            var city = new City(settings.GeneratePopulation(new[]
                        {
                            "Timofey",
                            "Леночка",
                            "Android",
                            "Дыня",
                            "Степан"
                        },
                        new Random(4))
                    .ToHashSet(),
                settings.CityName);
            var inter = new StringInterface();
            var game = new Game(settings, city, inter);
            game.StartGame().Wait();
            game.GetGameStatus().Should().Be(WinState.PsychoWins);
            Approvals.Verify(inter.GetGameProcess());
        }
    }
}