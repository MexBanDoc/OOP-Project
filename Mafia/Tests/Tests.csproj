<?xml version="1.0" encoding="utf-8"?>
<Project ToolsVersion="4.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
    <Import Project="..\packages\ApprovalTests.5.4.5\build\ApprovalTests.props" Condition="Exists('..\packages\ApprovalTests.5.4.5\build\ApprovalTests.props')" />
    <Import Project="..\packages\NUnit.3.12.0\build\NUnit.props" Condition="Exists('..\packages\NUnit.3.12.0\build\NUnit.props')" />
    <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" Condition="Exists('$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props')" />
    <PropertyGroup>
        <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
        <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
        <ProjectGuid>{BA3B5CD2-FDAC-40BF-8CA4-AE4559E45D70}</ProjectGuid>
        <ProjectTypeGuids>{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}</ProjectTypeGuids>
        <OutputType>Library</OutputType>
        <AppDesignerFolder>Properties</AppDesignerFolder>
        <RootNamespace>Tests</RootNamespace>
        <AssemblyName>Tests</AssemblyName>
        <TargetFrameworkVersion>v4.8</TargetFrameworkVersion>
        <FileAlignment>512</FileAlignment>
    </PropertyGroup>
    <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
        <PlatformTarget>AnyCPU</PlatformTarget>
        <DebugSymbols>true</DebugSymbols>
        <DebugType>full</DebugType>
        <Optimize>false</Optimize>
        <OutputPath>bin\Debug\</OutputPath>
        <DefineConstants>DEBUG;TRACE</DefineConstants>
        <ErrorReport>prompt</ErrorReport>
        <WarningLevel>4</WarningLevel>
    </PropertyGroup>
    <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
        <PlatformTarget>AnyCPU</PlatformTarget>
        <DebugType>pdbonly</DebugType>
        <Optimize>true</Optimize>
        <OutputPath>bin\Release\</OutputPath>
        <DefineConstants>TRACE</DefineConstants>
        <ErrorReport>prompt</ErrorReport>
        <WarningLevel>4</WarningLevel>
    </PropertyGroup>
    <ItemGroup>
        <Reference Include="ApprovalTests, Version=3.0.0.0, Culture=neutral, PublicKeyToken=11bd7d124fc62e0f">
          <HintPath>..\packages\ApprovalTests.5.4.5\lib\net461\ApprovalTests.dll</HintPath>
          <Private>True</Private>
        </Reference>
        <Reference Include="ApprovalUtilities, Version=3.0.0.0, Culture=neutral, PublicKeyToken=11bd7d124fc62e0f">
          <HintPath>..\packages\ApprovalUtilities.5.4.5\lib\net461\ApprovalUtilities.dll</HintPath>
          <Private>True</Private>
        </Reference>
        <Reference Include="Castle.Core, Version=4.0.0.0, Culture=neutral, PublicKeyToken=407dd0808d44fbdc">
          <HintPath>..\packages\Castle.Core.4.4.0\lib\net45\Castle.Core.dll</HintPath>
          <Private>True</Private>
        </Reference>
        <Reference Include="DiffEngine, Version=1.0.0.0, Culture=neutral, PublicKeyToken=c7a34512ecd69090">
          <HintPath>..\packages\DiffEngine.6.4.6\lib\net461\DiffEngine.dll</HintPath>
          <Private>True</Private>
        </Reference>
        <Reference Include="EmptyFiles, Version=1.0.0.0, Culture=neutral, PublicKeyToken=c7a34512ecd69090">
          <HintPath>..\packages\EmptyFiles.2.3.1\lib\net461\EmptyFiles.dll</HintPath>
          <Private>True</Private>
        </Reference>
        <Reference Include="FluentAssertions, Version=5.10.3.0, Culture=neutral, PublicKeyToken=33f2691a05b67b6a">
          <HintPath>..\packages\FluentAssertions.5.10.3\lib\net47\FluentAssertions.dll</HintPath>
          <Private>True</Private>
        </Reference>
        <Reference Include="Microsoft.Extensions.DependencyInjection.Abstractions, Version=3.1.4.0, Culture=neutral, PublicKeyToken=adb9793829ddae60">
          <HintPath>..\packages\Microsoft.Extensions.DependencyInjection.Abstractions.3.1.4\lib\netstandard2.0\Microsoft.Extensions.DependencyInjection.Abstractions.dll</HintPath>
          <Private>True</Private>
        </Reference>
        <Reference Include="Microsoft.Win32.Registry, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
          <HintPath>..\packages\Microsoft.Win32.Registry.5.0.0\lib\net461\Microsoft.Win32.Registry.dll</HintPath>
          <Private>True</Private>
        </Reference>
        <Reference Include="Moq, Version=4.16.0.0, Culture=neutral, PublicKeyToken=69f491c39445e920">
          <HintPath>..\packages\Moq.4.16.0\lib\net45\Moq.dll</HintPath>
          <Private>True</Private>
        </Reference>
        <Reference Include="mscorlib" />
        <Reference Include="nunit.framework, Version=3.12.0.0, Culture=neutral, PublicKeyToken=2638cd05610744eb">
          <HintPath>..\packages\NUnit.3.12.0\lib\net45\nunit.framework.dll</HintPath>
          <Private>True</Private>
        </Reference>
        <Reference Include="System" />
        <Reference Include="System.Configuration" />
        <Reference Include="System.Core" />
        <Reference Include="System.Data" />
        <Reference Include="System.Data.SqlClient, Version=4.6.1.2, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
          <HintPath>..\packages\System.Data.SqlClient.4.8.2\lib\net461\System.Data.SqlClient.dll</HintPath>
          <Private>True</Private>
        </Reference>
        <Reference Include="System.Management" />
        <Reference Include="System.Runtime.CompilerServices.Unsafe, Version=4.0.4.1, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
          <HintPath>..\packages\System.Runtime.CompilerServices.Unsafe.4.5.3\lib\net461\System.Runtime.CompilerServices.Unsafe.dll</HintPath>
          <Private>True</Private>
        </Reference>
        <Reference Include="System.Runtime.InteropServices.RuntimeInformation, Version=4.0.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
          <HintPath>..\packages\System.Runtime.InteropServices.RuntimeInformation.4.3.0\lib\net45\System.Runtime.InteropServices.RuntimeInformation.dll</HintPath>
          <Private>True</Private>
        </Reference>
        <Reference Include="System.Security.AccessControl, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
          <HintPath>..\packages\System.Security.AccessControl.5.0.0\lib\net461\System.Security.AccessControl.dll</HintPath>
          <Private>True</Private>
        </Reference>
        <Reference Include="System.Security.Principal.Windows, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
          <HintPath>..\packages\System.Security.Principal.Windows.5.0.0\lib\net461\System.Security.Principal.Windows.dll</HintPath>
          <Private>True</Private>
        </Reference>
        <Reference Include="System.Threading.Tasks.Extensions, Version=4.2.0.1, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51">
          <HintPath>..\packages\System.Threading.Tasks.Extensions.4.5.4\lib\net461\System.Threading.Tasks.Extensions.dll</HintPath>
          <Private>True</Private>
        </Reference>
        <Reference Include="System.Xml" />
        <Reference Include="System.Xml.Linq" />
        <Reference Include="TextCopy, Version=1.0.0.0, Culture=neutral, PublicKeyToken=1ca091877d12ca03">
          <HintPath>..\packages\TextCopy.4.2.0\lib\net461\TextCopy.dll</HintPath>
          <Private>True</Private>
        </Reference>
    </ItemGroup>
    <ItemGroup>
        <Compile Include="IntegrationTests\Tests.cs" />
        <Compile Include="TestApp\TestTGBot.cs" />
        <Compile Include="TestDomain\TestCity.cs" />
        <Compile Include="TestDomain\TestGame.cs" />
        <Compile Include="TestDomain\TestInteractions.cs" />
        <Compile Include="TestDomain\TestSettings.cs" />
        <Compile Include="TestInfrastucture\TestLinq.cs" />
        <Compile Include="Properties\AssemblyInfo.cs" />
    </ItemGroup>
    <ItemGroup>
      <ProjectReference Include="..\Mafia.csproj">
        <Project>{ef0d95f8-3e13-4811-8ad7-5d91415f6706}</Project>
        <Name>Mafia</Name>
      </ProjectReference>
    </ItemGroup>
    <ItemGroup>
      <None Include="packages.config" />
    </ItemGroup>
    <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
    <Target Name="EnsureNuGetPackageBuildImports" BeforeTargets="PrepareForBuild">
      <PropertyGroup>
        <ErrorText>This project references NuGet package(s) that are missing on this computer. Enable NuGet Package Restore to download them.  For more information, see http://go.microsoft.com/fwlink/?LinkID=322105.The missing file is {0}.</ErrorText>
      </PropertyGroup>
      <Error Condition="!Exists('..\packages\NUnit.3.12.0\build\NUnit.props')" Text="$([System.String]::Format('$(ErrorText)', '..\packages\NUnit.3.12.0\build\NUnit.props'))" />
      <Error Condition="!Exists('..\packages\EmptyFiles.2.3.1\build\EmptyFiles.targets')" Text="$([System.String]::Format('$(ErrorText)', '..\packages\EmptyFiles.2.3.1\build\EmptyFiles.targets'))" />
      <Error Condition="!Exists('..\packages\ApprovalTests.5.4.5\build\ApprovalTests.props')" Text="$([System.String]::Format('$(ErrorText)', '..\packages\ApprovalTests.5.4.5\build\ApprovalTests.props'))" />
      <Error Condition="!Exists('..\packages\ApprovalTests.5.4.5\build\ApprovalTests.targets')" Text="$([System.String]::Format('$(ErrorText)', '..\packages\ApprovalTests.5.4.5\build\ApprovalTests.targets'))" />
    </Target>
    <Import Project="..\packages\EmptyFiles.2.3.1\build\EmptyFiles.targets" Condition="Exists('..\packages\EmptyFiles.2.3.1\build\EmptyFiles.targets')" />
    <Import Project="..\packages\ApprovalTests.5.4.5\build\ApprovalTests.targets" Condition="Exists('..\packages\ApprovalTests.5.4.5\build\ApprovalTests.targets')" />
    <!-- To modify your build process, add your task inside one of the targets below and uncomment it. 
         Other similar extension points exist, see Microsoft.Common.targets.
    <Target Name="BeforeBuild">
    </Target>
    <Target Name="AfterBuild">
    </Target>
    -->

</Project>
