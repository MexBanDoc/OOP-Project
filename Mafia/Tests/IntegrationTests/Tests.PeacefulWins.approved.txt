Город засыпает
Просыпается Русская Мафия Timofey
ЛенОЧКА Killed

Город просыпается
Просыпается Peaceful Timofey
Просыпается Peaceful Android
Просыпается Peaceful Дыня
Просыпается Peaceful God
Просыпается Peaceful Степан
Просыпается Peaceful Николай
Timofey Killed

Мирные победили
